### ISSUE TITLE
Spurious failure on harness from s2n-quic

### ISSUE THREAD

**Issue by zhassan-aws** [model-checking/kani#2650]
This bug was reported by @camshaft on [s2n-quic](https://github.com/aws/s2n-quic). I've extracted a stand-alone example for easy debugging:

```rust
use std::ops::{Deref, DerefMut};

//
// from = [[aaa][bbbbb][cc][ddddd]]
// to   = [[xx][xxx][xxxxx][xxxxxxxxx]]
//
// result:
// to'  = [[aa][abb][bbbcc][ddddddxxx]]
//
pub fn vectored_copy<A, B>(from: &[A], to: &mut [B])
where
    A: Deref<Target = [u8]>,
    B: Deref<Target = [u8]> + DerefMut,
{
    let mut from_index = 0;
    let mut from_offset = 0;

    let mut to_index = 0;
    let mut to_offset = 0;

    // iterate until we reach one of the ends
    while from_index < from.len() && to_index < to.len() {
        let from = unsafe {
            // Safety: this length is already checked in the while condition
            debug_assert!(from.len() > from_index);
            from.get_unchecked(from_index)
        };

        let to = unsafe {
            // Safety: this length is already checked in the while condition
            debug_assert!(to.len() > to_index);
            to.get_unchecked_mut(to_index)
        };

        {
            // copy the bytes in the current views
            let from = unsafe {
                // Safety: the slice offsets are checked at the end of the while loop
                debug_assert!(from.len() >= from_offset);
                from.get_unchecked(from_offset..)
            };

            let to = unsafe {
                // Safety: the slice offsets are checked at the end of the while loop
                debug_assert!(to.len() >= to_offset);
                to.get_unchecked_mut(to_offset..)
            };

            let len = from.len().min(to.len());

            unsafe {
                // Safety: by using the min of the two lengths we will never exceed
                //         either slice's buffer
                debug_assert!(from.len() >= len);
                debug_assert!(to.len() >= len);
                to.get_unchecked_mut(..len)
                    .copy_from_slice(from.get_unchecked(..len));
            }

            // increment the offsets
            from_offset += len;
            to_offset += len;
        }

        // check if the `from` is done
        if from.len() == from_offset {
            from_index += 1;
            from_offset = 0;
        }

        // check if the `to` is done
        if to.len() == to_offset {
            to_index += 1;
            to_offset = 0;
        }
    }
}

struct F {
    values: [u8; 2],
    len: usize,
}

impl std::ops::Deref for F {
    type Target = [u8];

    fn deref(&self) -> &Self::Target {
        &self.values[..self.len]
    }
}

impl std::ops::DerefMut for F {
    fn deref_mut(&mut self) -> &mut Self::Target {
        &mut self.values[..self.len]
    }
}

struct G {
    values: [F; 2],
    len: usize,
}

impl std::ops::Deref for G {
    type Target = [F];

    fn deref(&self) -> &Self::Target {
        &self.values[..self.len]
    }
}

impl std::ops::DerefMut for G {
    fn deref_mut(&mut self) -> &mut Self::Target {
        &mut self.values[..self.len]
    }
}


#[kani::proof]
#[kani::unwind(5)]
fn main() {
    let ndlen = kani::any_where(|l| *l <= 2);
    let from = G { values: [F { values: [255, 255], len: ndlen }, F { values: [255, 255], len: 2}], len: 2};
    let mut to = G { values: [F { values: [255, 255], len: 1 }, F { values: [255, 255], len: 1}], len: 2};
    vectored_copy(&from, &mut to);
}
```

using the following command line invocation:

```
cargo kani
```

with Kani version: a2a1e855046

I expected to see this happen: Verification successful

Instead, this happened: Verification failed:
```
Check 16: core::slice::index::slice_end_index_len_fail_rt.assertion.1
         - Status: FAILURE
         - Description: "This is a placeholder message; Kani doesn't support message formatted at runtime"
         - Location: ../../.rustup/toolchains/nightly-2023-07-01-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/core/src/slice/index.rs:76:5 in function core::slice::index::slice_end_index_len_fail_rt
```
Running concrete playback using the following command:
```
$ cargo kani --enable-unstable --concrete-playback=inplace
```
generates the following test:
```rust
#[test]
fn kani_concrete_playback_main_11688673437834003895() {
    let concrete_vals: Vec<Vec<u8>> = vec![
        // 2ul
        vec![2, 0, 0, 0, 0, 0, 0, 0],
    ];
    kani::concrete_playback_run(concrete_vals, main);
}
```
If I run this test, it doesn't fail:
```
$ cargo kani playback -Zconcrete-playback 
Kani Rust Verifier 0.33.0 (cargo plugin)
   Compiling deref v0.1.0 (/home/ubuntu/examples/deref)
    Finished test [unoptimized + debuginfo] target(s) in 0.65s
     Running unittests src/main.rs (target/x86_64-unknown-linux-gnu/debug/deps/deref-f52e207434ba7027)

running 1 test
test kani_concrete_playback_main_11688673437834003895 ... ok

test result: ok. 1 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out; finished in 0.00s
```

Also, if I replace `ndlen` in the harness with 2 as the playback test suggests and run Kani, verification succeeds:
```
SUMMARY:
 ** 0 of 96 failed (4 unreachable)

VERIFICATION:- SUCCESSFUL
Verification Time: 0.68793035s

Complete - 1 successfully verified harnesses, 0 failures, 1 total.
```

**Comment by remi-delmas-3000** [comment]
Hi, I went as far back as kani 0.17.0 and cbmc 5.72.0 and the problem persists.

looking at the trace it appears that the `self.len` for  is not constrained as expected

```Rust
fn vectored_copy(from: &[F], to: &mut [F])
{
    // *_index variables keep track of the outer slice index
    // *_offset variables keep track of the innder slice index
    let mut from_index = 0;
    let mut from_offset = 0;

    let mut to_index = 0;
    let mut to_offset = 0;

    // iterate until we reach one of the ends
    while from_index < from.len() && to_index < to.len() {

        let f_from: &F = unsafe {
            // Safety: this length is already checked in the while condition
            debug_assert!(from_index < from.len());
            from.get_unchecked(from_index)
        };
        // Safety: this length is already checked in the while condition
        debug_assert!(to_index < to.len());
        // passes
        assert!(f_from.len <= 2);

        let f_to: &mut F = unsafe {
            to.get_unchecked_mut(to_index)
        };
        // fails here with
        let _dummy_from_index = from_index; // 1
        let _dummy_from_offset = from_offset; // 0
        let _dummy_to_index = to_index; // 1
        let _dummy_to_offset = to_offset; // 0
        let _dummy_f_to_len = f_to.len; // 18374686479671623680ul
        assert!(f_to.len <= 2);
        // fixes the problem downstream
        kani::assume(f_to.len <= 2);

 ```

Looking at the SSA encoding produced by CBMC, it seems like at some point the variable encoding the return value of `get_unchecked_mut`  is only constrained to be equal to itself, instead of being equal to the actual expression that defines the return value.

```
// these constraints encode the body of get_unchecked_mut for the type F
{-4782} _RNvXs0_NtNtCsdR6DzVNQbQk_4core5slice5indexjINtB5_10SliceIndexSNtCshFivlfURAdq_18s2n_quick_spurious1FE17get_unchecked_mutB10_::1::var_1::self!0@2#1 = _RINvMNtCsdR6DzVNQbQk_4core5sliceSNtCshFivlfURAdq_18s2n_quick_spurious1F17get_unchecked_mutjEBx_::1::var_2::index!0@2#1
{-4783} _RNvXs0_NtNtCsdR6DzVNQbQk_4core5slice5indexjINtB5_10SliceIndexSNtCshFivlfURAdq_18s2n_quick_spurious1FE17get_unchecked_mutB10_::1::var_2::slice!0@2#1..data = cast(address_of(main::1::var_12::to!0@1), struct tag-F*)
{-4784} _RNvXs0_NtNtCsdR6DzVNQbQk_4core5slice5indexjINtB5_10SliceIndexSNtCshFivlfURAdq_18s2n_quick_spurious1FE17get_unchecked_mutB10_::1::var_2::slice!0@2#1..len = 2
{-4785} _RNvXs0_NtNtCsdR6DzVNQbQk_4core5slice5indexjINtB5_10SliceIndexSNtCshFivlfURAdq_18s2n_quick_spurious1FE17get_unchecked_mutB10_::1::var_3!0@2#2 = 0
{-4786} _RNvMs_NtNtCsdR6DzVNQbQk_4core3ptr7mut_ptrOSNtCshFivlfURAdq_18s2n_quick_spurious1F10as_mut_ptrBH_::1::var_1::self!0@3#1..data = cast(address_of(main::1::var_12::to!0@1), struct tag-F*)
{-4787} _RNvMs_NtNtCsdR6DzVNQbQk_4core3ptr7mut_ptrOSNtCshFivlfURAdq_18s2n_quick_spurious1F10as_mut_ptrBH_::1::var_1::self!0@3#1..len = 2
{-4788} _RNvMs_NtNtCsdR6DzVNQbQk_4core3ptr7mut_ptrOSNtCshFivlfURAdq_18s2n_quick_spurious1F10as_mut_ptrBH_::1::var_0!0@3#2 = cast(address_of(main::1::var_12::to!0@1), struct tag-F*)
{-4789} goto_symex::return_value::_RNvMs_NtNtCsdR6DzVNQbQk_4core3ptr7mut_ptrOSNtCshFivlfURAdq_18s2n_quick_spurious1F10as_mut_ptrBH_!0#3 = cast(address_of(main::1::var_12::to!0@1), struct tag-F*)
{-4790} _RNvXs0_NtNtCsdR6DzVNQbQk_4core5slice5indexjINtB5_10SliceIndexSNtCshFivlfURAdq_18s2n_quick_spurious1FE17get_unchecked_mutB10_::1::var_6!0@2#2 = cast(address_of(main::1::var_12::to!0@1), struct tag-F*)
{-4791} _RNvMNtNtCsdR6DzVNQbQk_4core3ptr7mut_ptrONtCshFivlfURAdq_18s2n_quick_spurious1F3addBE_::1::var_1::self!0@3#1 = cast(address_of(main::1::var_12::to!0@1), struct tag-F*)
{-4792} _RNvMNtNtCsdR6DzVNQbQk_4core3ptr7mut_ptrONtCshFivlfURAdq_18s2n_quick_spurious1F3addBE_::1::var_2::count!0@3#1 = _RNvXs0_NtNtCsdR6DzVNQbQk_4core5slice5indexjINtB5_10SliceIndexSNtCshFivlfURAdq_18s2n_quick_spurious1FE17get_unchecked_mutB10_::1::var_1::self!0@2#1
{-4793} ¬goto_symex::\guard#15 ⇒ ¬(overflow-*(cast(_RNvMNtNtCsdR6DzVNQbQk_4core3ptr7mut_ptrONtCshFivlfURAdq_18s2n_quick_spurious1F3addBE_::1::var_2::count!0@3#1, signedbv[64]), 8))
{-4794} ¬goto_symex::\guard#15 ⇒ ¬(overflow-+(cast(cast(address_of(main::1::var_12::to!0@1), unsignedbv[64]), signedbv[64]), cast(_RNvMNtNtCsdR6DzVNQbQk_4core3ptr7mut_ptrONtCshFivlfURAdq_18s2n_quick_spurious1F3addBE_::1::var_2::count!0@3#1, signedbv[64]) * 8))
{-4795} _RNvMNtNtCsdR6DzVNQbQk_4core3ptr7mut_ptrONtCshFivlfURAdq_18s2n_quick_spurious1F3addBE_::$tmp::tmp_statement_expression!0@3#2 = cast(address_of(main::1::var_12::to!0@1), struct tag-F*) + _RNvMNtNtCsdR6DzVNQbQk_4core3ptr7mut_ptrONtCshFivlfURAdq_18s2n_quick_spurious1F3addBE_::1::var_2::count!0@3#1
{-4796} _RNvMNtNtCsdR6DzVNQbQk_4core3ptr7mut_ptrONtCshFivlfURAdq_18s2n_quick_spurious1F3addBE_::1::var_0!0@3#2 = _RNvMNtNtCsdR6DzVNQbQk_4core3ptr7mut_ptrONtCshFivlfURAdq_18s2n_quick_spurious1F3addBE_::$tmp::tmp_statement_expression!0@3#2
{-4797} goto_symex::return_value::_RNvMNtNtCsdR6DzVNQbQk_4core3ptr7mut_ptrONtCshFivlfURAdq_18s2n_quick_spurious1F3addBE_!0#3 = _RNvMNtNtCsdR6DzVNQbQk_4core3ptr7mut_ptrONtCshFivlfURAdq_18s2n_quick_spurious1F3addBE_::1::var_0!0@3#2
{-4798} _RNvXs0_NtNtCsdR6DzVNQbQk_4core5slice5indexjINtB5_10SliceIndexSNtCshFivlfURAdq_18s2n_quick_spurious1FE17get_unchecked_mutB10_::1::var_0!0@2#2 = goto_symex::return_value::_RNvMNtNtCsdR6DzVNQbQk_4core3ptr7mut_ptrONtCshFivlfURAdq_18s2n_quick_spurious1F3addBE_!0#3

// following back from _RNvCshFivlfURAdq_18s2n_quick_spurious13vectored_copy::1::var_36::f_to!0@1#3 (the pointer from which the len field is extracted) 
// we arrive here, the return variable of the get_unchecked_mut for the type F (which returns an F*)
// we see that the constraints above encode the body of the function but the return value is unconstrained.

{-4799} goto_symex::return_value::_RNvXs0_NtNtCsdR6DzVNQbQk_4core5slice5indexjINtB5_10SliceIndexSNtCshFivlfURAdq_18s2n_quick_spurious1FE17get_unchecked_mutB10_!0#2 =
        goto_symex::return_value::_RNvXs0_NtNtCsdR6DzVNQbQk_4core5slice5indexjINtB5_10SliceIndexSNtCshFivlfURAdq_18s2n_quick_spurious1FE17get_unchecked_mutB10_!0#2


{-4800} _RINvMNtCsdR6DzVNQbQk_4core5sliceSNtCshFivlfURAdq_18s2n_quick_spurious1F17get_unchecked_mutjEBx_::1::var_3!0@2#2 = goto_symex::return_value::_RNvXs0_NtNtCsdR6DzVNQbQk_4core5slice5indexjINtB5_10SliceIndexSNtCshFivlfURAdq_18s2n_quick_spurious1FE17get_unchecked_mutB10_!0#2
{-4801} _RINvMNtCsdR6DzVNQbQk_4core5sliceSNtCshFivlfURAdq_18s2n_quick_spurious1F17get_unchecked_mutjEBx_::1::var_0!0@2#2 = _RINvMNtCsdR6DzVNQbQk_4core5sliceSNtCshFivlfURAdq_18s2n_quick_spurious1F17get_unchecked_mutjEBx_::1::var_3!0@2#2
{-4802} goto_symex::return_value::_RINvMNtCsdR6DzVNQbQk_4core5sliceSNtCshFivlfURAdq_18s2n_quick_spurious1F17get_unchecked_mutjEBx_!0#2 = _RINvMNtCsdR6DzVNQbQk_4core5sliceSNtCshFivlfURAdq_18s2n_quick_spurious1F17get_unchecked_mutjEBx_::1::var_0!0@2#2
{-4803} _RNvCshFivlfURAdq_18s2n_quick_spurious13vectored_copy::1::var_37!0@1#3 = goto_symex::return_value::_RINvMNtCsdR6DzVNQbQk_4core5sliceSNtCshFivlfURAdq_18s2n_quick_spurious1F17get_unchecked_mutjEBx_!0#2
{-4804} _RNvCshFivlfURAdq_18s2n_quick_spurious13vectored_copy::1::var_36::f_to!0@1#3 = _RNvCshFivlfURAdq_18s2n_quick_spurious13vectored_copy::1::var_37!0@1#3
{-4805} _RNvCshFivlfURAdq_18s2n_quick_spurious13vectored_copy::1::var_39::_dummy_from_index!0@1#3 = _RNvCshFivlfURAdq_18s2n_quick_spurious13vectored_copy::1::var_3::from_index!0@1#4
{-4806} _RNvCshFivlfURAdq_18s2n_quick_spurious13vectored_copy::1::var_40::_dummy_from_offset!0@1#3 = _RNvCshFivlfURAdq_18s2n_quick_spurious13vectored_copy::1::var_4::from_offset!0@1#5
{-4807} _RNvCshFivlfURAdq_18s2n_quick_spurious13vectored_copy::1::var_41::_dummy_to_index!0@1#3 = _RNvCshFivlfURAdq_18s2n_quick_spurious13vectored_copy::1::var_5::to_index!0@1#4
{-4808} _RNvCshFivlfURAdq_18s2n_quick_spurious13vectored_copy::1::var_42::_dummy_to_offset!0@1#3 = _RNvCshFivlfURAdq_18s2n_quick_spurious13vectored_copy::1::var_6::to_offset!0@1#5
{-4809} _RNvCshFivlfURAdq_18s2n_quick_spurious13vectored_copy::1::var_43::_dummy_f_to_len!0@1#3 = byte_extract_little_endian({ { { { main::1::var_12::to!0@1#5..values[[0]]..values[[0]], main::1::var_12::to!0@1#5..values[[0]]..values[[1]] }, main::1::var_12::to!0@1#5..values[[0]]..$pad1, main::1::var_12::to!0@1#5..values[[0]]..len }, { { main::1::var_12::to!0@1#5..values[[1]]..values[[0]], main::1::var_12::to!0@1#5..values[[1]]..values[[1]] }, main::1::var_12::to!0@1#5..values[[1]]..$pad1, main::1::var_12::to!0@1#5..values[[1]]..len } }, main::1::var_12::to!0@1#5..len }, pointer_offset(_RNvCshFivlfURAdq_18s2n_quick_spurious13vectored_copy::1::var_36::f_to!0@1#3) + 8, unsignedbv[64])
{-4810} _RNvCshFivlfURAdq_18s2n_quick_spurious13vectored_copy::1::var_48!0@1#3 = 

// _RNvCshFivlfURAdq_18s2n_quick_spurious13vectored_copy::1::var_36::f_to!0@1#3 is the result if to.get_unchecked(to_index), it points within the to : &mut G and 
// reassembled G struct
 byte_extract_little_endian(
    {
         // array of two F
        { 
            { { main::1::var_12::to!0@1#5..values[[0]]..values[[0]], main::1::var_12::to!0@1#5..values[[0]]..values[[1]] }, main::1::var_12::to!0@1#5..values[[0]]..$pad1, main::1::var_12::to!0@1#5..values[[0]]..len },
            { { main::1::var_12::to!0@1#5..values[[1]]..values[[0]], main::1::var_12::to!0@1#5..values[[1]]..values[[1]] }, main::1::var_12::to!0@1#5..values[[1]]..$pad1, main::1::var_12::to!0@1#5..values[[1]]..len }
        },
        // len of the array
        main::1::var_12::to!0@1#5..len
    }, pointer_offset(_RNvCshFivlfURAdq_18s2n_quick_spurious13vectored_copy::1::var_36::f_to!0@1#3) + 8 /* this must be the offset of the len field in an F struct */, unsignedbv[64]
)

{-4811} _RNvCshFivlfURAdq_18s2n_quick_spurious13vectored_copy::1::var_47!0@1#3 = (_RNvCshFivlfURAdq_18s2n_quick_spurious13vectored_copy::1::var_48!0@1#3 ≥ 3 ? 0 : 1)
{-4812} _RNvCshFivlfURAdq_18s2n_quick_spurious13vectored_copy::1::var_46!0@1#3 = (_RNvCshFivlfURAdq_18s2n_quick_spurious13vectored_copy::1::var_47!0@1#3 = 0 ? 1 : 0)
{-4813} _RNvCshFivlfURAdq_18s2n_quick_spurious13vectored_copy::1::var_45!0@1#3 = (_RNvCshFivlfURAdq_18s2n_quick_spurious13vectored_copy::1::var_46!0@1#3 = 0 ? 1 : 0)
{-4814} _RNvCshFivlfURAdq_18s2n_quick_spurious13vectored_copy::1::temp_3!0@2#2 ⇔ (_RNvCshFivlfURAdq_18s2n_quick_spurious13vectored_copy::1::var_45!0@1#3 ≠ 0)
├──────────────────────────
{1} ¬goto_symex::\guard#15 ⇒ _RNvCshFivlfURAdq_18s2n_quick_spurious13vectored_copy::1::temp_3!0@2#2
```
This looks like a CBMC bug, 

@tautschnig can you take a second look ?

**Comment by zhassan-aws** [comment]
A slightly reduced version:
```rust
use std::ops::{Deref, DerefMut};

pub fn vectored_copy<B>(from: &[&[u8]], to: &mut [B])
where
    B: Deref<Target = [u8]> + DerefMut,
{
    let mut from_index = 0;
    let mut from_offset = 0;

    let mut to_index = 0;
    let mut to_offset = 0;

    // iterate until we reach one of the ends
    while from_index < from.len() && to_index < to.len() {
        let from = unsafe {
            from.get_unchecked(from_index)
        };

        let to = unsafe {
            to.get_unchecked_mut(to_index)
        };

        {
            // copy the bytes in the current views
            let from = unsafe {
                from.get_unchecked(from_offset..)
            };

            let to = unsafe {
                to.get_unchecked_mut(to_offset..)
            };

            let len = from.len().min(to.len());

            unsafe {
                to.get_unchecked_mut(..len)
                    .copy_from_slice(from.get_unchecked(..len));
            }

            // increment the offsets
            from_offset += len;
            to_offset += len;
        }

        // check if the `from` is done
        if from.len() == from_offset {
            from_index += 1;
            from_offset = 0;
        }

        // check if the `to` is done
        if to.len() == to_offset {
            to_index += 1;
            to_offset = 0;
        }
    }
}

struct F {
    values: [u8; 2],
    len: usize,
}

impl std::ops::Deref for F {
    type Target = [u8];

    fn deref(&self) -> &Self::Target {
        &self.values[..self.len]
    }
}

impl std::ops::DerefMut for F {
    fn deref_mut(&mut self) -> &mut Self::Target {
        &mut self.values[..self.len]
    }
}

struct G {
    values: [F; 2],
    _len: usize,
}

impl std::ops::Deref for G {
    type Target = [F];

    fn deref(&self) -> &Self::Target {
        &self.values
    }
}

impl std::ops::DerefMut for G {
    fn deref_mut(&mut self) -> &mut Self::Target {
        &mut self.values
    }
}


#[kani::proof]
#[kani::unwind(5)]
fn main() {
    let ndlen = kani::any_where(|l| *l == 2);
    let from1 = [255, 255];
    let from2 = [255, 255];
    let from = [&from1[..ndlen], &from2[..]];
    let mut to = G { values: [F { values: [255, 255], len: 1 }, F { values: [255, 255], len: 1}], _len: 2};
    vectored_copy(&from, &mut to);
}
```

**Comment by remi-delmas-3000** [comment]
@tautschnig full reproducer posted above

**Comment by tautschnig** [comment]
This is caused by a bug in CBMC's handling of byte updates of non-constant width (which are generated from `to.get_unchecked_mut(..len).copy_from_slice(from.get_unchecked(..len));`). Fixed once https://github.com/diffblue/cbmc/pull/7855 is merged and part of a release.

**Comment by zhassan-aws** [comment]
diffblue/cbmc#7855 was merged. This issue should get fixed once we get a new CBMC release.
